File: app/src/main/java/rbsoftware/friendstagram/ui/activity/CreatePostActivity.java
package rbsoftware.friendstagram.ui.activity;

import android.app.ProgressDialog;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.app.NavUtils;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.Map;

import rbsoftware.friendstagram.ImageSelectListener;
import rbsoftware.friendstagram.R;
import rbsoftware.friendstagram.ui.fragment.SelectFilterFragment;
import rbsoftware.friendstagram.ui.fragment.SelectImageFragment;
import rbsoftware.friendstagram.ui.fragment.SharePostFragment;
import rbsoftware.friendstagram.model.Error;
import rbsoftware.friendstagram.model.Post;
import rbsoftware.friendstagram.model.Response;
import rbsoftware.friendstagram.service.AuthenticationService;
import rbsoftware.friendstagram.service.ImageService;
import rbsoftware.friendstagram.service.NetworkService;
import rbsoftware.friendstagram.service.PostsService;
import retrofit2.Call;
import retrofit2.Callback;

public class CreatePostActivity extends AppCompatActivity implements ImageSelectListener {

    private static final String TAG = "CreatePostActivity";
    private Uri imageURI;
    private String caption;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_post);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        getSupportActionBar().setDisplayShowTitleEnabled(false);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close);

        showImageSelectFragment();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_create_post, menu);
        return imageURI != null;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.item_next:
                nextClicked();
                return true;
        }
        return false;
    }

    @Override
    public void onBackPressed() {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.container);
        if (fragment instanceof SelectImageFragment) {
            // Workaround to fix issue 5
            NavUtils.navigateUpFromSameTask(this);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public void onImageSelected(Uri uri) {
        imageURI = uri;
        invalidateOptionsMenu();
    }

    private void nextClicked() {
        Fragment fragment = getSupportFragmentManager().findFragmentById(R.id.container);
        if (fragment instanceof SelectImageFragment) {
            showFilterSelectFragment();
        } else if (fragment instanceof SelectFilterFragment) {
            showSharePostFragment();
        } else if (fragment instanceof SharePostFragment) {
            if (isValidForm(fragment)) {
                uploadImage();
            }
        }
    }

    private boolean isValidForm(Fragment fragment) {
        assert fragment instanceof SharePostFragment;
        SharePostFragment postFragment = (SharePostFragment) fragment;
        if (TextUtils.isEmpty(postFragment.getCaption())) {
            postFragment.showRequiredCaption();
            return false;
        }
        caption = postFragment.getCaption();
        return true;
    }

    private void uploadImage() {
        final ProgressDialog dialog = ProgressDialog.show(this, "", "Uploading image...");
        try {
            FileInputStream inputStream = new FileInputStream(new File(new URI(imageURI.toString().replace(" ", "%20"))));
            ImageService.getInstance().uploadImage(inputStream, AuthenticationService.getInstance().getUsername(), new ImageService.ImageResponseHandler<Map>() {
                @Override
                public void onComplete(Map response) {
                    if (response.containsKey("exception")) {
                        Toast.makeText(getApplicationContext(), "Failed to upload image!", Toast.LENGTH_SHORT).show();
                        Log.e(TAG, "nextClicked: Failed to upload image", (IOException) response.get("exception"));
                    } else {
                        Log.d(TAG, "onComplete: Image upload successful!");
                        final String publicURL = (String) response.get("secure_url"); // ID to reference image in Cloudinary
                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                createPost(publicURL, caption);
                            }
                        });
                    }
                    dialog.dismiss();
                }
            });
        } catch (IOException e) {
            Toast.makeText(getApplicationContext(), "Failed to upload image!", Toast.LENGTH_SHORT).show();
            Log.e(TAG, "nextClicked: Failed to upload image", e);
            dialog.dismiss();
        } catch (URISyntaxException e) {
            Toast.makeText(getApplicationContext(), "An error occurred!", Toast.LENGTH_SHORT).show();
            Log.e(TAG, "uploadImage: Invalid URI", e);
            dialog.dismiss();
        }
    }

    private void createPost(String imageURL, String caption) {
        final ProgressDialog dialog = ProgressDialog.show(this, "", "Sharing post...");
        Call<Response<Post>> createPostTask = PostsService.getInstance().getAPI().createPost(new Post(imageURL, caption, new ArrayList<String>()));
        createPostTask.enqueue(new Callback<Response<Post>>() {
            @Override
            public void onResponse(Call<Response<Post>> call, retrofit2.Response<Response<Post>> response) {
                if (response.isSuccessful()) {
                    Log.d(TAG, "onResponse: Image upload successful");
                    Toast.makeText(getApplicationContext(), getString(R.string.success_image_upload), Toast.LENGTH_SHORT).show();
                    NavUtils.navigateUpFromSameTask(CreatePostActivity.this);
                } else {
                    Error error = NetworkService.parseError(response);
                    //TODO: Handle error
                }
                dialog.dismiss();
            }

            @Override
            public void onFailure(Call<Response<Post>> call, Throwable t) {
                Log.e(TAG, "onFailure: Failed to create post", t);
                dialog.dismiss();
                Toast.makeText(getApplicationContext(), getString(R.string.error_occurred), Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void showImageSelectFragment() {
        showFragment(SelectImageFragment.newInstance(), false);
    }

    private void showFilterSelectFragment() {
        assert imageURI != null;
        showFragment(SelectFilterFragment.newInstance(imageURI), true);
    }

    private void showSharePostFragment() {
        assert imageURI != null;
        showFragment(SharePostFragment.newInstance(imageURI), true);
    }

    private void showFragment(Fragment fragment, boolean addToStack) {
        FragmentManager manager = getSupportFragmentManager();
        FragmentTransaction transaction = manager.beginTransaction();

        transaction.replace(R.id.container, fragment);
        if (addToStack)
            transaction.addToBackStack(null);
        transaction.commit();
    }
}

File: app/src/main/java/rbsoftware/friendstagram/ui/activity/CreatePostActivity.kt
package rbsoftware.friendstagram.ui.activity

import android.annotation.SuppressLint
import android.app.ProgressDialog
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import android.support.v4.graphics.drawable.DrawableCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.text.TextUtils
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.layout_alt_toolbar.*
import rbsoftware.friendstagram.InitializerApp
import rbsoftware.friendstagram.R
import rbsoftware.friendstagram.model.Post
import rbsoftware.friendstagram.service.AuthenticationService
import rbsoftware.friendstagram.service.NetworkService
import rbsoftware.friendstagram.showFragment
import rbsoftware.friendstagram.ui.fragment.SelectFilterFragment
import rbsoftware.friendstagram.ui.fragment.SelectImageFragment
import rbsoftware.friendstagram.ui.fragment.SharePostFragment
import rbsoftware.friendstagram.viewmodel.PostViewModel
import java.io.File
import java.io.FileInputStream
import java.net.URI
import java.net.URISyntaxException

/**
 * Created by Rushil on 8/18/2017.
 */
class CreatePostActivity : AppCompatActivity() {
    private val uiSubscriptions: CompositeDisposable = CompositeDisposable()

    private var imageURI: Uri? = null
    private var caption: String? = null
    private lateinit var authService: AuthenticationService
    private lateinit var postViewModel: PostViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_create_post)

        val closeIndicator = ContextCompat.getDrawable(this, R.drawable.ic_close)

        val daggerComponent = InitializerApp.servicesComponent
        daggerComponent.inject(this)
        authService = daggerComponent.authService()
        postViewModel = daggerComponent.postViewModel()

        setSupportActionBar(toolbar)
        DrawableCompat.setTint(closeIndicator, Color.WHITE)
        supportActionBar?.setDisplayShowTitleEnabled(false)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeAsUpIndicator(closeIndicator)
    }

    override fun onPostResume() {
        super.onPostResume()
        showImageSelectFragment()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_create_post, menu)
        return imageURI != null
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.item_next -> {
                nextClicked()
                return true
            }
        }
        return false
    }

    private fun nextClicked() {
        val fragment = supportFragmentManager.findFragmentById(R.id.container)
        if (fragment is SelectImageFragment) {
            showFilterSelectFragment()
        } else if (fragment is SelectFilterFragment) {
            showSharePostFragment()
        } else if (fragment is SharePostFragment) {
            if (isValidForm(fragment)) {
                uploadImage()
            }
        }
    }

    @SuppressLint("RxLeakedSubscription")
    private fun showImageSelectFragment() {
        val fragment = SelectImageFragment.newInstance()
        fragment.isAdapterInitialized()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.trampoline())
                .subscribe({
                    fragment.getOnImageSelected()?.let { subject ->
                        uiSubscriptions.add(
                                subject.observeOn(AndroidSchedulers.mainThread())
                                        .subscribeOn(Schedulers.trampoline())
                                        .subscribe({ uri ->
                                            imageURI = uri
                                            invalidateOptionsMenu()
                                        }, { e ->
                                            Log.e(TAG, getString(R.string.error_occurred), e)
                                        })
                        )
                    }
                }, this::onSubscriptionError)
        showFragment(fragment)
    }

    private fun showFilterSelectFragment() {
        imageURI?.let { showFragment(SelectFilterFragment.newInstance(it), true, "SelectFilter") }
    }

    private fun showSharePostFragment() {
        imageURI?.let { showFragment(SharePostFragment.newInstance(it), true, "SharePost") }
    }

    private fun isValidForm(fragment: Fragment): Boolean {
        if (fragment is SharePostFragment) {
            val sharePostFragment: SharePostFragment = fragment
            if (TextUtils.isEmpty(sharePostFragment.getCaption())) {
                sharePostFragment.showRequiredCaption()
                return false
            }
            caption = sharePostFragment.getCaption()
            return true
        }
        throw IllegalArgumentException("Fragment does not have form!")
    }

    @SuppressLint("RxLeakedSubscription")
    private fun uploadImage() {
        val dialog = ProgressDialog.show(this, "", "Uploading image...")
        try {
            val inputStream = FileInputStream(File(URI(imageURI.toString().replace(" ", "%20"))))
            postViewModel.uploadImage(inputStream, authService.username)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeOn(Schedulers.io())
                    .subscribe({ response ->
                        val publicURL = response["secure_url"] as String
                        caption?.let { createPost(publicURL, it) }
                        dialog.dismiss()
                    }, { err ->
                        Toast.makeText(applicationContext, getString(R.string.error_image_upload), Toast.LENGTH_SHORT).show()
                        Log.e(TAG, "Failed to upload image", err)
                        dialog.dismiss()
                    })
        } catch (e: URISyntaxException) {
            Toast.makeText(applicationContext, getString(R.string.error_occurred), Toast.LENGTH_SHORT).show()
            Log.e(TAG, "uploadImage: Invalid URI", e)
            dialog.dismiss()
        }
    }

    @SuppressLint("RxLeakedSubscription")
    private fun createPost(imageURL: String, caption: String) {
        val dialog = ProgressDialog.show(this, "", "Sharing post...")
        postViewModel.createPost(Post(imageURL, caption))
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe({ response ->
                    if (response.isSuccessful) {
                        Toast.makeText(applicationContext, getString(R.string.success_image_upload), Toast.LENGTH_SHORT).show()
                        dialog.dismiss()
                        finish()
                    } else {
                        val error: String? = NetworkService.parseError(response.errorBody())
                        handleError(error)
                        dialog.dismiss()
                    }
                }, this::onNetworkError)
    }

    private fun handleError(error: String?) {
        when {
            error == null -> Toast.makeText(applicationContext, getString(R.string.error_occurred), Toast.LENGTH_SHORT).show()
            error == "bad token" || error.contains("User ID is Null") -> {
                Toast.makeText(applicationContext, getString(R.string.error_login_invalid), Toast.LENGTH_LONG).show()
                authService.logout()

                val intent = Intent(this, LoginRegisterActivity::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                startActivity(intent)
                finish()
            }
        }
    }

    private fun onNetworkError(error: Throwable) {
        Log.e(TAG, "Network errors", error)
        Toast.makeText(applicationContext, getString(R.string.error_occurred), Toast.LENGTH_SHORT).show()
    }

    private fun onSubscriptionError(error: Throwable) {
        throw RuntimeException(error)
    }

    companion object {
        private const val TAG = "CreatePostActivity"
    }
}